src/tox/session/env_select.py

```diff
@@ -0,0 +1,289 @@
+from __future__ import annotations
+
+from argparse import ArgumentParser
+from collections import Counter
+from dataclasses import dataclass
+from itertools import chain
+from typing import TYPE_CHECKING, Any, Iterable, Iterator, List, cast
+
+from tox.config.loader.str_convert import StrConvert
+from tox.tox_env.api import ToxEnvCreateArgs
+from tox.tox_env.register import REGISTER
+from tox.tox_env.runner import RunToxEnv
+
+from ..config.loader.memory import MemoryLoader
+from ..report import HandledError
+from ..tox_env.errors import Skip
+from ..tox_env.package import PackageToxEnv
+
+if TYPE_CHECKING:
+    from tox.session.state import State
+
+
+class CliEnv:
+    """CLI tox env selection"""
+
+    def __init__(self, value: None | list[str] | str = None):
+        if isinstance(value, str):
+            value = StrConvert().to(value, of_type=List[str], factory=None)
+        self._names: list[str] | None = value
+
+    def __iter__(self) -> Iterator[str]:
+        if not self.is_all and self._names is not None:  # pragma: no branch
+            yield from self._names
+
+    def __str__(self) -> str:
+        return "ALL" if self.is_all else ("<env_list>" if self.is_default_list else ",".join(self))
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({'' if self.is_default_list else repr(str(self))})"
+
+    def __eq__(self, other: Any) -> bool:
+        return type(self) == type(other) and self._names == other._names
+
+    def __ne__(self, other: Any) -> bool:
+        return not (self == other)
+
+    @property
+    def is_all(self) -> bool:
+        return self._names is not None and "ALL" in self._names
+
+    @property
+    def is_default_list(self) -> bool:
+        return not (self._names or [])
+
+
+def register_env_select_flags(
+    parser: ArgumentParser,
+    default: CliEnv | None,
+    multiple: bool = True,
+    group_only: bool = False,
+) -> None:
+    if not multiple:
+        parser.add_argument("-e", dest="env", help="environment to run", default=default, type=CliEnv)
+        return
+
+    if not group_only:
+        group = parser.add_argument_group("select target environment(s)")
+        excl = group.add_mutually_exclusive_group(required=False)
+        help_msg = "enumerate (ALL -> all environments, not set -> use <env_list> from config)"
+        excl.add_argument("-e", dest="env", help=help_msg, default=default, type=CliEnv)
+        # excl.add_argument("-m", dest="labels", metavar="label", help="labels to run", default=[], type=str, nargs="+")
+        # excl.add_argument("-f", dest="factors", help="factors to run", default=[], type=str, nargs="+")
```

Comment:
leftovers?

Reply:
More like to do.  The current changes just refactor the world to make factor and tag base selection possible.  I should have made this PR a draft, sorry 😁